runtime/minor_gc.c

```diff
@@ -546,7 +545,7 @@ void caml_empty_minor_heap_promote(caml_domain_state* domain,
 
       for( r = ref_start ; r < foreign_major_ref->ptr && r < ref_end ; r++ )
       {
-        oldify_one (&st, **r, *r);
+        oldify_one (&st, *((volatile value *) *r), *r);
```

Comment:
Am I understanding correctly that the problem is the read of the third argument and second argument could end up being different due to the race?
If that's right, then maybe it's clearer to use an intermediate value, .e.g.

```c
value * pr = *r;
oldify_one (&st, *pr,pr);
```

Reply:
The problem is not that the reads can end up being different: if `*pr` is a pointer and another thread has promoted it, reading the old value here is not an issue as it now points to a forwarding block, which will be detected by `oldify_one`. The problem is that the race is between a `volatile` write and a plain load, namely, `*pr` in your proposal. To avoid undefined behaviour, in principle, we would need to use relaxed atomic accesses. But we have [decided](https://github.com/ocaml/ocaml/pull/12473) that `volatile` is to be used as an equivalent of relaxed accesses in some cases in the runtime; hence my proposal of making the read `volatile`.

I agree that using an intermediate value is more legible. How about:
```c
volatile value * pr = *r;
oldify_one (&st, *pr, pr);
```

Reply:
Looks good! Thanks for the explanation.